lib/gc/Transforms/LowerTileVectorPass.cpp

```diff
@@ -0,0 +1,668 @@
+    return transferReadOp;
+  SmallVector<OpFoldResult> mixedSourceDims =
+      tensor::getMixedSizes(builder, loc, source);
+  Value mask =
+      builder.create<vector::CreateMaskOp>(loc, maskType, mixedSourceDims);
+  return vector::maskOperation(builder, transferReadOp, mask)->getResult(0);
+}
+
+/// Vectorize a `tensor::expandshape` to these 3 Ops:
+///   Vector::TransferReadOp - Reads a vector from the source tensor
+///   ShapeCastOp - Reshape the data based on the target.
+///   vector::TransferWriteOp. - Write the result vector back to the destination
+///   tensor
+template <class T>
+LogicalResult lowerTensorReshapeOp(RewriterBase &rewriter, Operation *inputOp,
+                                   SmallVectorImpl<Value> &newResults,
+                                   ArrayRef<int64_t> inputVectorSizes) {
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(inputOp);
+
+  auto src = inputOp->getOperand(0);
+  auto srcType = cast<ShapedType>(src.getType());
+  OpResult result = inputOp->getResults()[0];
+  auto resultType = cast<ShapedType>(result.getType());
+  ArrayRef<int64_t> resultShape = resultType.getShape();
+  ArrayRef<int64_t> srcShape = srcType.getShape();
+  Location loc = inputOp->getLoc();
+
+  SmallVector<int64_t> srcVectorizedShape(srcType.getRank());
+  llvm::SmallDenseMap<int64_t, int64_t> shapeScales;
+  ReshapeVectorizeHelper reshapeHelper(srcVectorizedShape, shapeScales,
+                                       resultShape, srcShape);
+
+  srcVectorizedShape.assign(srcShape.begin(), srcShape.end());
+  if (!inputVectorSizes.empty()) {
+    reshapeHelper.getScalesDim(inputVectorSizes);
+    getReshapeOperationVectorizeShape<T>(reshapeHelper);
+  }
+  // generate read operation
+  auto padValue = rewriter.create<arith::ConstantOp>(
+      loc, rewriter.getZeroAttr(srcType.getElementType()));
+  Value readResult = vector::createReadOrMaskedRead(
+      rewriter, loc, src,
+      inputVectorSizes.empty() ? srcType.getShape() : srcVectorizedShape,
+      padValue, true);
+
+  auto shapeCastType =
+      VectorType::get(inputVectorSizes.empty() ? resultShape : inputVectorSizes,
+                      resultType.getElementType());
+  vector::ShapeCastOp shapeCastOp =
+      rewriter.create<vector::ShapeCastOp>(loc, shapeCastType, readResult);
+
+  // generate write operation
+  SmallVector<OpFoldResult> destSizes(resultShape.size());
+  llvm::transform(resultShape, std::begin(destSizes), [&rewriter](size_t size) {
+    return rewriter.getIndexAttr(size);
+  });
+
+  Operation *write = createWriteOrMaskedWrite(
+      rewriter, loc, shapeCastOp->getResults()[0], destSizes,
+      inputVectorSizes.empty() ? resultShape : inputVectorSizes, true);
+  newResults.push_back(write->getResult(0));
+  return success();
+}
+
+/// Vectorize a `tensor::concat` to these 3 Ops:
+///   Tensor::EmptyOp - The result tensor.
+///   Vector::TransferWriteOp - Write the result vector back to the destination
```

Comment:
Should this be a TransferRead op?

Reply:
No. This means that we write the results into the final large concat tensor.

Reply:
Before writing, is there a reading?

Reply:
Thanks~  Add new comments~